add a worker queue to get updates from the main results thread

s-hertel opened the pull request:
##### SUMMARY
Create a multiprocessing queue for each WorkerProcess so plugins can get user input from the main results thread rather than use sys.stdin directly.
 
 `pause` and `paramiko_ssh` now use the queue, and the connection plugin's stdin attr is deprecated to get a little closer to being able spawn a new process without copying the existing one.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

s-hertel committed:
Create a queue per WorkerProcess for intra-task updates from the main results thread

Update pause to use the worker queue

s-hertel committed:
Deprecate ConnectionBase()._new_stdin

s-hertel committed:
Rename method to prompt_until and add msg argument that is displayed before reading stdin

Call display.prompt_until instead of importing send_prompt, delete send_prompt from worker_sync

Add a note about why input() isn't used for the first pause prompt if seconds is None

Fix continue condition (none) for no custom prompt + a timer - wait for timeout or user-interrupt

Delete positional arg instead of deprecating it since it's a helper class for paramiko_ssh and nothing should be using it, and remove extra locking

nitzmahone commented on lib/ansible/plugins/connection/paramiko_ssh.py:
```diff
@@ -250,6 +251,24 @@
 SETTINGS_REGEX = re.compile(r'(\w+)(?:\s*=\s*|\s+)(.+)')
 
 
+def _deprecate_arg0(func):
```
Maybe I'm missing something, but doing this with a non-generic decorator seems like an overly complex way to just do an inline arg deprecation on `MyAddPolicy.__init__`:
```
def __init__(self, new_stdin, connection):
    if new_stdin:
        display.deprecated(
            "MyAddPolicy.__init__() got deprecated argument: 'new_stdin'. "
            "Instantiate the class with the connection plugin as the only argument instead, "
            "and prompting will occur via the worker queue.",
            version='2.19'
        )
    ...
```

For that matter, the deprecation itself seems pretty generous- I assume we're worried about things that have subclassed our plugin?

s-hertel replied:
We'd need to turn new_stdin and connection into a kwargs instead, whereas I was deprecating passing it at all. This was just for 3rd party plugins potentially using it, in an abundance of caution. I'll just delete arg 0 if this seems like overkill - paramiko isn't using this now and since this a helper class to paramiko, if anything else is using this, it should have not imo.

s-hertel replied:
I've removed the arg altogether for simplicity.

nitzmahone commented on lib/ansible/plugins/connection/paramiko_ssh.py:
```diff
@@ -279,14 +300,21 @@ def missing_host_key(self, client, hostname, key):
-            tcflush(sys.stdin, TCIFLUSH)
+                inp = input(AUTHENTICITY_MSG % (hostname, ktype, fingerprint))
+                sys.stdin = old_stdin
+            else:
```
The original code was already doing this "wrong" IMO anyway- a blind asynchronous call to Python `input` is fraught with peril in a world where `display` expects roughly exclusive control of stdout/stderr. Letting `display` lose the context that we're doing a prompt by separating the output from the input with two calls and a relative eternity for things to change from underneath us between multiple threads/processes and a bunch of IPC isn't going to improve the situation :laughing: .

I've actually got a working way to just fully remote the entire display object from the controller into the worker (and vice versa for any arbitrary objects) without an external `Manager`, but I don't think we're ready to do that just yet.

With a fairly small change, we can just remote a call to the existing native `display.prompt` on the controller side, similarly to what we're doing with `display.display` now. You'd just do like `display.display` did and delegate to a worker-local queue-backed implementation of `display.prompt` that uses the dataclass you've already created (maybe with some slight tweaks) and sends it to the controller over the queue (including the message to display) and blocks on the queue for its response and returns/raises appropriately. While the worker is waiting for its result, once the controller has dequeued the PromptSend, it just calls the real `display.prompt` with the values you passed over, then sends the response back through the worker's queue. This mechanism doesn't scale that well, but for how infrequently these kinds of prompts happen, it's fine, and the whole thing can go away once we've got arbitrary bidrectional object remoting anyway. This way, the plugin code can still just directly call the appropriate methods on `display`, which is much clearer anyway, and noone's the wiser when we change how that's actually occurring, whether it's over a queue, a remoted `Display` object, or a carrier pigeon.

Remember too, that even though there's a lock to prevent >1 connection from doing this concurrently, there's potentially a *bunch* of unrelated stuff happening concurrently that's talking to `Display` while we're waiting for someone to answer these prompts, so its state can change in all sorts of problematic ways. If we really want this to be robust, we'll also need to synchronize or queue writes to `display` on the controller, such that a pending prompt will prevent anything else from writing to stdout. That part might be "bonus points" for this PR, since from what I can see, we don't do much (if anything) to prevent that right now anyway, so this change doesn't make that any worse than it already is. If we did that, though, we could actually remove the `connection_lock/unlock` stuff entirely, since the plugin will just naturally block until the controller gets back to it, and the controller's `Display` class is the one arbitrating all the actual concurrency, state of stdout, and listening on stdin for one prompt at a time.

s-hertel replied:
I think I've addressed this in the latest non-documentation commit. I renamed the method to `prompt_until` since it doesn't just read now, and pass a message as the first argument. I'm returning the value from the queue after sending the prompt, so plugins can just use the display method.

I thought I would be able to refactor `prompt_until` so paramiko_ssh could still call `input()` via the queue (and maybe one of the corner cases for `pause` - if there is no timeout and we're waiting for enter), but I didn't realize KeyboardInterrupt (which `input()` can raise), needs to be handled in the main thread (as opposed the daemon results thread). Here's a simple example https://gist.github.com/s-hertel/ebd57439ee74d0e40028ccb5fd6a2d78 to visualize the issue + possible fix. If the exception handling needs to occur in the main thread, then the interrupt/done logic would also need to be moved there as well, so I left a comment about why `self.prompt` isn't used when it seems like it should be for the default use-case.

I'm not sure I've addressed your last point, but I have now manually tested the paramiko_ssh prompt against a dozen hosts, and it is working with free/linear/host_pinned without the connection_lock/unlock, so I removed it. (The display lock *is* used in `prompt_until` since I setup/cleanup for each prompt.)

nitzmahone commented on lib/ansible/utils/display.py:
```diff
@@ -18,31 +18,45 @@
 from __future__ import (absolute_import, division, print_function)
 __metaclass__ = type
 
+try:
```
So I'm thinking that all of this can go away if we do the remoting stuff I mentioned above and just delegate to the existing `prompt` method, since IIRC it does everything we care about already anyway via `input` (which is totally fine if we synchronize the access to it).

nitzmahone commented on lib/ansible/utils/display.py:
```diff
@@ -520,7 +596,122 @@ def do_var_prompt(self, varname, private=True, prompt=None, encrypt=None, confir
             tty_size = 0
         self.columns = max(79, tty_size - 1)
+
+    def do_non_blocking_read_until(
```
possibly ditto here if we can just use the normal blocking `input`

nitzmahone reviewed (changes requested):
I think we can drastically simplify this in a more future-proof way pretty easily- let's kick that around next week...

nitzmahone commented on lib/ansible/plugins/connection/__init__.py:
```diff
@@ -88,6 +89,16 @@ def __init__(self, play_context, new_stdin, shell=None, *args, **kwargs):
 
+    @property
+    def _new_stdin(self):
+        display.deprecated(
```
also, if we want to, we could keep this working with a `BytesIO` stdin simulator, but I'm kinda leaning toward "not worth the trouble"

s-hertel replied:
I think it would be an interesting exercise/fun but probably need more time. I could remove the deprecation if we want to do that.

nitzmahone replied:
IIRC the only other consumer of that is `synchronize`, which as you know "there be dragons" :laughing: :dragon:

s-hertel replied:
For simplicity, I've decided not to do this - `synchronize` uses it to instantiate a new local connection (and the arg is now optional), so `synchronize` will just be able to delete it.

s-hertel committed:
changelog

s-hertel committed:
add coverage for the pause cornercase I just broke and unbroke

s-hertel committed:
default new_stdin to None since it is not required

s-hertel committed:
put weird default back in pause

undo unintended change - removed remaining pause duration following user input

mkrizek commented:
!waffling needs_revision

s-hertel committed:
fix ignoring non-interrupt input if a pause duration is specified

make cosmetic change to tailor interface to pause (single character comparison)

use documentation option formatting and clarify about prompting with a timeout

add a test

s-hertel committed:
just guessing...

s-hertel committed:
Instantiate the worker's intermediate results queue in __init__

remove unused file

nitzmahone commented on lib/ansible/executor/task_queue_manager.py:
```diff
@@ -64,6 +66,21 @@ def __init__(self, *args, **kwargs):
+    interrupt_input: t.Iterable[bytes] = None
+    complete_input: t.Iterable[bytes] = None
+
+    kwargs: t.Mapping = field(init=False, default_factory=dict)
```
This and the `__post_init__` shouldn't be necessary, just grab `(instance).__dict__` at the call-site in `prompt_until` to get the instance's values as a splattable dict- there are a couple other ways with dataclasses helper methods as well, but `__dict__` is easy for this case.

s-hertel replied:
@nitzmahone The `worker_id` attr isn't a kwarg, so I can pop it out of the `__dict__` to get the kwargs. I'd probably want to shallow copy first so I'm not mutating it. But if you want to get rid of this, my first inclination would have been just to use the public attributes, no need to have kwargs defined on the dataclass at all, it was just convenient and repr()ed nicely. Want me to copy and use `instance.__dict__` anyway?

To be specific I think I prefer the current implementation, followed by

<details><summary>using public attrs</summary>

```diff
diff --git a/lib/ansible/executor/task_queue_manager.py b/lib/ansible/executor/task_queue_manager.py
index 2345bddf4c..b14ab6303c 100644
--- a/lib/ansible/executor/task_queue_manager.py
+++ b/lib/ansible/executor/task_queue_manager.py
@@ -75,11 +75,6 @@ class PromptSend:
     interrupt_input: t.Iterable[bytes] = None
     complete_input: t.Iterable[bytes] = None
 
-    kwargs: t.Mapping = field(init=False, default_factory=dict)
-
-    def __post_init__(self):
-        self.kwargs = dict(private=self.private, seconds=self.seconds, interrupt_input=self.interrupt_input, complete_input=self.complete_input)
-
 
 class FinalQueue(multiprocessing.queues.Queue):
     def __init__(self, *args, **kwargs):
diff --git a/lib/ansible/plugins/strategy/__init__.py b/lib/ansible/plugins/strategy/__init__.py
index 2c5810e727..cce387b57a 100644
--- a/lib/ansible/plugins/strategy/__init__.py
+++ b/lib/ansible/plugins/strategy/__init__.py
@@ -129,7 +129,13 @@ def results_thread_main(strategy):
                     strategy._results.append(result)
             elif isinstance(result, PromptSend):
                 try:
-                    value = display.prompt_until(result.prompt, **result.kwargs)
+                    value = display.prompt_until(
+                        result.prompt,
+                        private=result.private,
+                        seconds=result.seconds,
+                        complete_input=result.complete_input,
+                        interrupt_input=result.interrupt_input,
+                    )
                 except AnsibleError as e:
                     value = e
                 except BaseException as e:
```

</details>

followed by

<details><summary>using __dict__ </summary>

```diff
diff --git a/lib/ansible/executor/task_queue_manager.py b/lib/ansible/executor/task_queue_manager.py
index 2345bddf4c..b14ab6303c 100644
--- a/lib/ansible/executor/task_queue_manager.py
+++ b/lib/ansible/executor/task_queue_manager.py
@@ -75,11 +75,6 @@ class PromptSend:
     interrupt_input: t.Iterable[bytes] = None
     complete_input: t.Iterable[bytes] = None
 
-    kwargs: t.Mapping = field(init=False, default_factory=dict)
-
-    def __post_init__(self):
-        self.kwargs = dict(private=self.private, seconds=self.seconds, interrupt_input=self.interrupt_input, complete_input=self.complete_input)
-
 
 class FinalQueue(multiprocessing.queues.Queue):
     def __init__(self, *args, **kwargs):
diff --git a/lib/ansible/plugins/strategy/__init__.py b/lib/ansible/plugins/strategy/__init__.py
index 2c5810e727..e7b7a9b8bd 100644
--- a/lib/ansible/plugins/strategy/__init__.py
+++ b/lib/ansible/plugins/strategy/__init__.py
@@ -128,8 +128,11 @@ def results_thread_main(strategy):
                 with strategy._results_lock:
                     strategy._results.append(result)
             elif isinstance(result, PromptSend):
+                prompt_kwargs = result.__dict__.copy()
+                del prompt_kwargs['worker_id']
+                prompt = prompt_kwargs.pop('prompt')
                 try:
-                    value = display.prompt_until(result.prompt, **result.kwargs)
+                    value = display.prompt_until(prompt, **prompt_kwargs)
                 except AnsibleError as e:
                     value = e
                 except BaseException as e:
```

</details>

nitzmahone replied:
Yeah, thinking about it a little more, this is just one of those reflexive design red flags that actually doesn't hurt in this specific case (it's a really brittle way to call a quasi-related method in the general case). OTOH, since there's only one caller, it's probably safer and simpler to just use the public attrs directly at the call-site- the end result is net less code anyway and one less thing you have to worry about keeping in sync.

s-hertel replied:
I updated it to use the public attrs at the call site.

nitzmahone commented on lib/ansible/plugins/action/pause.py:
```diff
@@ -168,143 +102,67 @@ def run(self, tmp=None, task_vars=None):
-                    intr = b'\x03'  # value for Ctrl+C
+                if seconds is not None:
+                    # setup the alarm handler
+                    signal.signal(signal.SIGALRM, timeout_handler)
```
If we're always remoting to the controller `prompt_until` on the timeout cases anyway and `pause` is already a run-once type action, do we need this `SIGALRM` stuff in the action at all anymore?

s-hertel replied:
I removed it.

nitzmahone reviewed (changes requested):
Looking great- noted a couple of minor things to tweak, but I think we're just about there... Still pretty sure we could eliminate the raw mode stuff altogether, but we can do it later.

s-hertel committed:
remove signal handling, helped catch timeout bugs in the controller but is not necessary

s-hertel committed:
use public attributes for simplicity

s-hertel committed:
only call curses.setupterm() when an initial prompt occurs

patch curses unit tests since importing display doesn't call curses.setupterm()

s-hertel commented on lib/ansible/utils/display.py:
```diff
@@ -244,6 +237,20 @@ def setup_prompt(stdin_fd, stdout_fd, seconds, echo):
+    else:
+        global MOVE_TO_BOL
+        global CLEAR_TO_EOL
+        MOVE_TO_BOL = curses.tigetstr('cr') or MOVE_TO_BOL
```
```suggestion
        # curses.tigetstr() returns None in some circumstances
        MOVE_TO_BOL = curses.tigetstr('cr') or MOVE_TO_BOL
```

nitzmahone commented on lib/ansible/utils/display.py:
```diff
@@ -520,7 +591,139 @@ def do_var_prompt(self, varname, private=True, prompt=None, encrypt=None, confir
+            setupterm()
+            self.setup_curses = True
+
+        if (
```
Don't we need to check `not HAS_CURSES` here as well?

s-hertel replied:
Why? If curses couldn't be imported or setupterm fails, we'll use the hardcoded values for MOVE_TO_BOL and CLEAR_TO_EOL.

nitzmahone replied:
:facepalm: Yep, you're right- I forgot the original code was only actually using curses to query the terminfo and always doing the terminal state management manually (bleh). So it's consistent with the original code as you have it.

nitzmahone reviewed (changes requested):
One little tweak and I think we're there.

s-hertel committed:
simplify unit test
